lib/active_publisher/async/in_memory_adapter/async_queue.rb

```diff
@@ -2,30 +2,44 @@ module ActivePublisher
   module Async
     module InMemoryAdapter
       class AsyncQueue
+        # These strategies are used to determine what to do with messages when the queue is full.
+        # :raise - Raise an error and drop the message.
+        # :drop - Silently drop the message.
+        # :wait - Wait for space in the queue to become available.
+        BACK_PRESSURE_STRATEGIES = [:raise, :drop, :wait].freeze
+
         include ::ActivePublisher::Logging
 
-        attr_accessor :drop_messages_when_queue_full,
+        attr_accessor :back_pressure_strategy,
           :max_queue_size,
           :supervisor_interval
 
         attr_reader :consumer, :queue, :supervisor
 
-        def initialize(drop_messages_when_queue_full, max_queue_size, supervisor_interval)
-          @drop_messages_when_queue_full = drop_messages_when_queue_full
+        def initialize(back_pressure_strategy, max_queue_size, supervisor_interval)
+          self.back_pressure_strategy = back_pressure_strategy
           @max_queue_size = max_queue_size
           @supervisor_interval = supervisor_interval
           @queue = ::MultiOpQueue::Queue.new
           create_and_supervise_consumer!
         end
 
-        def push(message)
-          # default of 1_000_000 messages
-          if queue.size > max_queue_size
-            # Drop messages if the queue is full and we were configured to do so
-            return if drop_messages_when_queue_full
+        def back_pressure_strategy=(strategy)
+          fail ::ArgumentError, "Invalid back pressure strategy: #{strategy}" unless BACK_PRESSURE_STRATEGIES.include?(strategy)
+          @back_pressure_strategy = strategy
+        end
 
-            # By default we will raise an error to push the responsibility onto the caller
-            fail ::ActivePublisher::Async::InMemoryAdapter::UnableToPersistMessageError, "Queue is full, messages will be dropped."
+        def push(message)
+          if queue.size >= max_queue_size
+            case back_pressure_strategy
+            when :drop
+              return
+            when :raise
+              fail ::ActivePublisher::Async::InMemoryAdapter::UnableToPersistMessageError, "Queue is full, messages will be dropped."
+            when :wait
+              # This is a really crappy way to wait
+              sleep 0.01 until queue.size < max_queue_size
```

Comment:
I would swap out the Queue for a SizedQueue, but we're using @abrandoned 's MultiOpQueue which has a different interface for popping a batch of messages. I contemplated mocking that interface, but that took me to a dark place. I thought about protecting this with a mutex, but we're going to hit one in the queue and that seemed slow. Pick your poison, right? What do you think.

Reply:
I can your point about `sleep` being a code-smell, but since this is intended to be an edge-case for backpressure, I think it's pretty reasonably.  Maybe we should add in some activesupport instrumentation around the wait loop so we can track how often we hit this and how long we spend here. That might help provide some insight that eliminates the concern.